### PR TITLE
CLI Updates

### PR DESCRIPTION
**Pull request by stuartleeks**
# Work towards #2706 and #2707

## What is being addressed

Describe the current behavior you are modifying. Please also remember to update any impacted documentation.

## How is this addressed

- Add invoke action commands
- Add shell completion for shared service IDs
- Add basic table query for overall costs
- Add migrations command
- Add raw output option
- Add airlock request title
- airlock -> airlock-request and fix error text
- Fix airlock request id shell completion
- Add table format for airlock request submit response and review response
- Fix up airlock request review URL
- Add example airlock-request creation to README.md
- Update API: Handle missing app role name
- Add airlock-request cancel command
- Add workspace service delete command
- Update output command to take response and exit on error (for more consistent exit code handling)

### CODE REVIEW

File: cli/tre/api_client.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import sys
+from typing import Union
 import click
 import json
 import msal
 import os
 
 from httpx import Client, Response
+from httpx._types import QueryParamTypes
```
Comment (martinpeck) [inline cli/tre/api_client.py:9]: Looks like you're importing a "private" module. Is QueryParamTypes a stable API to depend upon?
Do you need to type it like this, or can it just be typed as a Dict?

File: cli/tre/commands/shared_services/shared_service.py [hunk 2]
```diff
@@ -10,8 +10,17 @@
 from .operations import shared_service_operations
 
 
+def shared_service_id_completion(ctx, param, incomplete):
```
Comment (martinpeck) [inline cli/tre/commands/shared_services/shared_service.py:13]: nit: This is fine....it's just that you've been type-hinting all the other methods, so I wondered "why not this one?"

File: cli/tre/api_client.py (hunk 1 above)
Comment (stuartleeks) [reply cli/tre/api_client.py]: I used that type as that is what the `params` property on the (public) `request` method is typed as.
 
 I've updated to a `dict` now

File: cli/tre/commands/shared_services/shared_service.py (hunk 2 above)
Comment (stuartleeks) [reply cli/tre/commands/shared_services/shared_service.py]: I've pushed a commit with type annotations added to the completion methods

**Review by martinpeck — APPROVED** [review]
LGTM!

File: api_app/services/aad_authentication.py [hunk 3]
```diff
@@ -265,7 +265,7 @@ def get_workspace_role_assignment_details(self, workspace: Workspace):
 
             if principal_type == "User" and principal_id in user_emails:
                 app_role_id = role_assignment["appRoleId"]
-                app_role_name = inverted_app_role_ids[app_role_id]
+                app_role_name = inverted_app_role_ids.get(app_role_id)
```
[the remark this answers is not in the document]
Comment (martinpeck) [reply api_app/services/aad_authentication.py:268]: Yep. Makes sense.

**Pull request merged**